File: walk_route.rb
# walk_route.rb
#!/usr/bin/env ruby
require 'json'
require 'net/http'
require 'uri'
require 'optparse'

DATA_FILE = 'walk_route.json'
DEFAULT_SPEED = 5.0
CALORIES_PER_KM = 55

class Waypoint
  attr_accessor :name, :lat, :lng

  def initialize(name, lat, lng)
    @name = name
    @lat = lat
    @lng = lng
  end

  def to_hash
    { name: @name, lat: @lat, lng: @lng }
  end

  def self.from_hash(h)
    new(h['name'], h['lat'], h['lng'])
  end
end

class RoutePlanner
  attr_reader :waypoints, :speed, :weight

  def initialize
    @waypoints = []
    @speed = DEFAULT_SPEED
    @weight = 70
    load
  end

  def load
    if File.exist?(DATA_FILE)
      data = JSON.parse(File.read(DATA_FILE))
      @waypoints = data['waypoints'].map { |h| Waypoint.from_hash(h) }
      @speed = data['speed'] || DEFAULT_SPEED
      @weight = data['weight'] || 70
    end
  end

  def save
    File.write(DATA_FILE, JSON.pretty_generate({
      waypoints: @waypoints.map(&:to_hash),
      speed: @speed,
      weight: @weight
    }))
  end

  def geocode(query)
    url = URI("https://nominatim.openstreetmap.org/search")
    params = { q: query, format: 'json', limit: 1 }
    url.query = URI.encode_www_form(params)
    req = Net::HTTP::Get.new(url)
    req['User-Agent'] = 'WalkRoutePlanner/1.0'
    res = Net::HTTP.start(url.host, url.port, use_ssl: true) { |http| http.request(req) }
    return nil unless res.is_a?(Net::HTTPSuccess)
    data = JSON.parse(res.body)
    return nil if data.empty?
    [data[0]['lat'].to_f, data[0]['lon'].to_f]
  end

  def add(location)
    if location.include?(',')
      parts = location.split(',').map(&:strip)
      if parts.size == 2
        lat = parts[0].to_f
        lng = parts[1].to_f
        if lat != 0.0 && lng != 0.0
          name = "(%.4f,%.4f)" % [lat, lng]
          @waypoints << Waypoint.new(name, lat, lng)
          save
          puts "✅ Added waypoint: #{name}"
          return
        end
      end
    end
    coords = geocode(location)
    unless coords
      puts "❌ Could not geocode '#{location}'"
      return
    end
    lat, lng = coords
    @waypoints << Waypoint.new(location, lat, lng)
    save
    puts "✅ Added waypoint: #{location} (%.4f, %.4f)" % [lat, lng]
  end

  def list
    if @waypoints.empty?
      puts "No waypoints."
      return
    end
    puts "\n📍 Waypoints:"
    @waypoints.each_with_index do |w, i|
      puts "#{i+1}: #{w.name} (%.4f, %.4f)" % [w.lat, w.lng]
    end
  end

  def remove(index)
    if index < 1 || index > @waypoints.size
      puts "❌ Invalid index. Must be between 1 and #{@waypoints.size}"
      return
    end
    removed = @waypoints.delete_at(index-1)
    save
    puts "✅ Removed waypoint: #{removed.name}"
  end

  def clear
    @waypoints = []
    save
    puts "✅ All waypoints cleared."
  end

  def speed(speed)
    @speed = speed
    save
    puts "✅ Walking speed set to #{'%.1f' % speed} km/h"
  end

  def route
    if @waypoints.size < 2
      puts "Need at least 2 waypoints to calculate a route."
      return
    end
    coords = @waypoints.map { |w| "#{w.lng},#{w.lat}" }.join(';')
    url = URI("http://router.project-osrm.org/route/v1/walking/#{coords}?overview=full&geometries=polyline&steps=true")
    res = Net::HTTP.get(url)
    data = JSON.parse(res)
    unless data['routes'] && !data['routes'].empty?
      puts "❌ No route found."
      return
    end
    route = data['routes'][0]
    distance_km = route['distance'] / 1000.0
    duration_min = route['duration'] / 60.0
    elevation = distance_km * 12
    calories = distance_km * CALORIES_PER_KM * (@weight / 70.0)

    puts "\n🚶 Walking Route:"
    puts "Total distance: %.2f km" % distance_km
    puts "Total duration: %.1f min (at %.1f km/h)" % [duration_min, @speed]
    puts "Elevation gain: ~%.0f m" % elevation
    puts "Calories burned: ~%.0f kcal" % calories

    if route['legs']
      puts "\nTurn-by-turn:"
      step_num = 1
      route['legs'].each do |leg|
        leg['steps'].each do |step|
          instruction = step['maneuver']['instruction']
          dist = step['distance'] / 1000.0
          puts "#{step_num}. #{instruction} (%.2f km)" % dist
          step_num += 1
        end
      end
    end
  end
end

options = {}
OptionParser.new do |opts|
  opts.banner = "Usage: walk_route.rb <command> [options]"
  opts.on("add LOCATION", "Add waypoint") { |v| options[:add] = v }
  opts.on("list", "List waypoints") { options[:list] = true }
  opts.on("route", "Calculate route") { options[:route] = true }
  opts.on("remove INDEX", Integer, "Remove waypoint") { |v| options[:remove] = v }
  opts.on("clear", "Clear all") { options[:clear] = true }
  opts.on("speed KMH", Float, "Set walking speed") { |v| options[:speed] = v }
end.parse!

planner = RoutePlanner.new
if options[:add]
  planner.add(options[:add])
elsif options[:list]
  planner.list
elsif options[:route]
  planner.route
elsif options[:remove]
  planner.remove(options[:remove])
elsif options[:clear]
  planner.clear
elsif options[:speed]
  planner.speed(options[:speed])
else
  puts "Usage: walk_route.rb [add|list|route|remove|clear|speed]"
end

File: walk_route.py
# walk_route.py
import sys
import os
import json
import requests
import argparse
import math
from datetime import datetime

DATA_FILE = "walk_route.json"
WALKING_SPEED = 5.0  # km/h
CALORIES_PER_KM = 55  # approx for a 70kg person

class Waypoint:
    def __init__(self, name, lat, lng):
        self.name = name
        self.lat = lat
        self.lng = lng

    def to_dict(self):
        return {"name": self.name, "lat": self.lat, "lng": self.lng}

    @classmethod
    def from_dict(cls, d):
        return cls(d["name"], d["lat"], d["lng"])

class RoutePlanner:
    def __init__(self):
        self.waypoints = []
        self.speed = WALKING_SPEED
        self.weight = 70  # kg
        self.load()

    def load(self):
        if os.path.exists(DATA_FILE):
            with open(DATA_FILE, "r") as f:
                data = json.load(f)
                self.waypoints = [Waypoint.from_dict(w) for w in data.get("waypoints", [])]
                self.speed = data.get("speed", WALKING_SPEED)
                self.weight = data.get("weight", 70)

    def save(self):
        with open(DATA_FILE, "w") as f:
            json.dump({
                "waypoints": [w.to_dict() for w in self.waypoints],
                "speed": self.speed,
                "weight": self.weight
            }, f, indent=2)

    def geocode(self, query):
        url = "https://nominatim.openstreetmap.org/search"
        params = {"q": query, "format": "json", "limit": 1}
        headers = {"User-Agent": "WalkRoutePlanner/1.0"}
        resp = requests.get(url, params=params, headers=headers, timeout=10)
        if resp.status_code != 200:
            return None
        data = resp.json()
        if not data:
            return None
        return float(data[0]["lat"]), float(data[0]["lon"])

    def add(self, location):
        if "," in location:
            try:
                parts = location.split(",")
                lat = float(parts[0].strip())
                lng = float(parts[1].strip())
                name = f"({lat:.4f},{lng:.4f})"
                self.waypoints.append(Waypoint(name, lat, lng))
                self.save()
                print(f"✅ Added waypoint: {name}")
                return
            except ValueError:
                pass
        coords = self.geocode(location)
        if not coords:
            print(f"❌ Could not geocode '{location}'")
            return
        lat, lng = coords
        self.waypoints.append(Waypoint(location, lat, lng))
        self.save()
        print(f"✅ Added waypoint: {location} ({lat:.4f}, {lng:.4f})")

    def list(self):
        if not self.waypoints:
            print("No waypoints.")
            return
        print("\n📍 Waypoints:")
        for i, w in enumerate(self.waypoints, 1):
            print(f"{i}: {w.name} ({w.lat:.4f}, {w.lng:.4f})")

    def remove(self, index):
        if index < 1 or index > len(self.waypoints):
            print(f"❌ Invalid index. Must be between 1 and {len(self.waypoints)}")
            return
        removed = self.waypoints.pop(index - 1)
        self.save()
        print(f"✅ Removed waypoint: {removed.name}")

    def clear(self):
        self.waypoints = []
        self.save()
        print("✅ All waypoints cleared.")

    def set_speed(self, speed):
        self.speed = speed
        self.save()
        print(f"✅ Walking speed set to {speed:.1f} km/h")

    def route(self):
        if len(self.waypoints) < 2:
            print("Need at least 2 waypoints to calculate a route.")
            return
        coords = ";".join(f"{w.lng},{w.lat}" for w in self.waypoints)
        url = f"http://router.project-osrm.org/route/v1/walking/{coords}"
        params = {"overview": "full", "geometries": "polyline", "steps": "true"}
        resp = requests.get(url, params=params, timeout=10)
        if resp.status_code != 200:
            print("❌ Error fetching route.")
            return
        data = resp.json()
        if "routes" not in data or not data["routes"]:
            print("❌ No route found.")
            return
        route = data["routes"][0]
        distance_km = route["distance"] / 1000
        duration_min = route["duration"] / 60
        # Estimate elevation gain (simplified)
        elevation_gain = self.estimate_elevation(route)
        # Calories burned
        calories = distance_km * CALORIES_PER_KM * (self.weight / 70)

        print("\n🚶 Walking Route:")
        print(f"Total distance: {distance_km:.2f} km")
        print(f"Total duration: {duration_min:.1f} min (at {self.speed:.1f} km/h)")
        print(f"Elevation gain: ~{elevation_gain:.0f} m")
        print(f"Calories burned: ~{calories:.0f} kcal")

        legs = route.get("legs", [])
        if legs:
            print("\nTurn-by-turn:")
            step_num = 1
            for leg in legs:
                for step in leg.get("steps", []):
                    instruction = step.get("maneuver", {}).get("instruction", "")
                    dist = step.get("distance", 0) / 1000
                    if instruction:
                        print(f"{step_num}. {instruction} ({dist:.2f} km)")
                        step_num += 1

    def estimate_elevation(self, route):
        # Simplified elevation gain from route
        # In a real implementation, we would use elevation data from DEM
        # For now, return a dummy value based on distance
        distance_km = route["distance"] / 1000
        return distance_km * 12  # rough approximation: 12m per km

def main():
    parser = argparse.ArgumentParser(description="Walking Route Planner")
    subparsers = parser.add_subparsers(dest="cmd", required=True)

    add_parser = subparsers.add_parser("add")
    add_parser.add_argument("location")

    subparsers.add_parser("list")
    subparsers.add_parser("route")

    remove_parser = subparsers.add_parser("remove")
    remove_parser.add_argument("index", type=int)

    subparsers.add_parser("clear")

    speed_parser = subparsers.add_parser("speed")
    speed_parser.add_argument("speed", type=float)

    args = parser.parse_args()
    planner = RoutePlanner()

    if args.cmd == "add":
        planner.add(args.location)
    elif args.cmd == "list":
        planner.list()
    elif args.cmd == "route":
        planner.route()
    elif args.cmd == "remove":
        planner.remove(args.index)
    elif args.cmd == "clear":
        planner.clear()
    elif args.cmd == "speed":
        planner.set_speed(args.speed)

if __name__ == "__main__":
    main()
